### DISCUSSION TITLE
SLIP over UART device

### DISCUSSION THREAD

**Discussion by ofirshe** [zephyrproject-rtos/zephyr#54203] [Q&A]
Hey, 

I have a designated UART device in my device tree and I want to run a TCP/IP application over this device. Is it possible as in Linux ? 
I saw such a driver called located in zephyr\drivers\net\slip.c, but there is no DT_DRV_COMPAT so how can I link it with my device ? Or maybe I have to create one by myself.

**Comment by jukkar** [comment]
Hi, the SLIP driver in the tree is mainly used for qemu<->host communication. I have heard people using it to connect from real device so it should be possible to do what you want. If you want to try this, please consider sending patches to upstream for any needed changes.

**Reply by ofirshe** [reply]
Hey @jukkar, 

Thank you very much for your answer. I would like to advice with you how should I approach this task. First,  I add a line in my custom dts file zephyr,uart-pipe = &lpuart1; in 'chosen' section and also enabled CONFIG_UART_PIPE=y in my project configuration file. Once enabling UART_PIPE, this leads to CONFIG_SLIP to be enabled which results in including zephyr\drivers\net\slip.c in the build.
In order to be able to use BSD networking api, I have to enable CONFIG_NETWORKING, CONFIG_NET_SOCKETS and 
CONFIG_NET_TCP also. When enabling CONFIG_NETWORKING, the build system starts searching for source files in zephyr\drivers\ethernet folder but no one found because it found in zephyr\drivers\net. Thus, I have a compilation warning:
CMake Warning at zephyr/CMakeLists.txt:824 (message):
  No SOURCES given to Zephyr library: drivers__ethernet

  Excluding target from build.


Should I replicate drivers\net\slip.c into zephyr\drivers\ethernet\my_custom_slip.c ? 
If I will do so, I'll have two slip drivers being compiled in the same build which will trigger NET_DEVICE_INIT twice, So I'm wondering if this is the right solution. 

Best regards, 
Ofir.

**Reply by jukkar** [reply]
By default TAP mode (Ethernet) is set when SLIP is enabled, this requires Ethernet driver support. But there is Ethernet driver setup in `drivers/net/slip.c` already when TAP is enabled, so you should not need anything else for the Ethernet to be supported.
I have not done anything like this, so don't know if your setup has something wrong. But you should not need to create a new Ethernet driver as slip.c has that already.

**Reply by ofirshe** [reply]
I'm fully agree with you, but in order to use all the BSD networking api, I must enable CONFIG_NETWORKING which cause the build mechanism to search for source files in zephyr\drivers\ethernet folder. I just want to get rid of this warning 
No SOURCES given to Zephyr library: drivers__ethernet

Excluding target from build.

Do you have any suggestions ?

**Reply by jukkar** [reply]
ping @pbtgit do you have any idea what might be wrong here, how did you fix you issue in #48697?

**Reply by ofirshe** [reply]
Thanks @jukkar

These are my configurations to be able to develop a simple TCP/IP application over SLIP:
CONFIG_SLIP=y
CONFIG_NET_DRIVERS=y
CONFIG_NETWORKING=y
CONFIG_NET_UDP=y
CONFIG_NET_TCP=y
CONFIG_NET_IPV4=y 
CONFIG_INIT_STACKS=y
CONFIG_NET_LOG=y
CONFIG_LOG=y
CONFIG_NET_CONFIG_SETTINGS=y
CONFIG_NET_CONFIG_NEED_IPV4=y
CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.1"
CONFIG_NET_SOCKETS=y
CONFIG_NET_SOCKETS_POSIX_NAMES=y
CONFIG_MAIN_STACK_SIZE=65536
CONFIG_NET_PKT_RX_COUNT=16
CONFIG_NET_PKT_TX_COUNT=16
CONFIG_NET_BUF_RX_COUNT=64
CONFIG_NET_BUF_TX_COUNT=64
CONFIG_NET_MAX_CONN=64

**Comment by jukkar** [comment]
See also https://github.com/zephyrproject-rtos/zephyr/discussions/48697

**Comment by pbtgit** [comment]
As for the issues in #48697  my conclusion was that it was due to the MCU on NRF boards - The UART over MCU  only worked stable with smaller packages than 64 bytes:  Using the pins to physical UART worked (with HW FC), and also when using the CDC UART.
(This holds for both Zephyr and Contiki)

**Reply by ofirshe** [reply]
@pbtgit Did you face this problem ? https://github.com/zephyrproject-rtos/zephyr/discussions/54203#discussioncomment-4843453****

**Reply by pbtgit** [reply]
Ah, No I don't have that problem - I wanted to have SLIP run in "TUN" mode rather than "TAP" mode.

**Reply by ofirshe** [reply]
@pbtgit Didn't you develop network application over SLIP ? 
In order to use all the BSD networking api, I must enable CONFIG_NETWORKING which cause the build mechanism to search for source files in zephyr\drivers\ethernet folder but we don't need to include in the build any file from this directory. I just want to get rid of this warning
No SOURCES given to Zephyr library: drivers__ethernet

**Reply by pbtgit** [reply]
Yes but in my scenario it could only use L3 (only IP frames ) so there was no need for Ethernet for running SLIP.
Essentially I think the following SLIP related settings differs from yours: 
CONFIG_NET_SLIP_TAP=n
CONFIG_SLIP_TAP=n
CONFIG_NET_L2_DUMMY=y

**Reply by ofirshe** [reply]
@pbtgit I see. So how should I solve this issue ? I don't want to replicate the zephyr\drivers\net\slip.c file into zephyr\drivers\ethernet directory. Do you have any suggestions or maybe open a bug for this ?

**Reply by pbtgit** [reply]
Maybe modify the build system so that zephyr\drivers\ethernet is skipped if TAP & SLIP already provides a driver in slip.c?

**Reply by ofirshe** [reply]
@pbtgit How can I do such a thing ? or maybe should I open an issue about it ?

**Reply by ofirshe** [reply]
@jukkar @pbtgit 
opened a bug for it https://github.com/zephyrproject-rtos/zephyr/issues/54460
